bvm/ballerina-core/src/main/java/org/ballerinalang/util/TableProvider.java

```diff
@@ -0,0 +1,310 @@
+*  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+*  KIND, either express or implied.  See the License for the
+*  specific language governing permissions and limitations
+*  under the License.
+*/
+package org.ballerinalang.util;
+
+import org.ballerinalang.model.types.BArrayType;
+import org.ballerinalang.model.types.BStructType;
+import org.ballerinalang.model.types.BTableType;
+import org.ballerinalang.model.types.BType;
+import org.ballerinalang.model.types.TypeTags;
+import org.ballerinalang.model.values.BBlobArray;
+import org.ballerinalang.model.values.BBooleanArray;
+import org.ballerinalang.model.values.BFloatArray;
+import org.ballerinalang.model.values.BIntArray;
+import org.ballerinalang.model.values.BStringArray;
+import org.ballerinalang.model.values.BStruct;
+import org.ballerinalang.model.values.BValue;
+import org.ballerinalang.util.exceptions.BallerinaException;
+
+import java.io.ByteArrayInputStream;
+import java.sql.Blob;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+/**
+ * {@code TableProvider} creates In Memory database for tables.
+ *
+ * @since 0.963.0
+ */
+public class TableProvider {
+
+    private static TableProvider tableProvider = null;
+    private int tableID;
+
+    private TableProvider() {
+        tableID = 0;
+    }
+
+    public static TableProvider getInstance() {
+        if (tableProvider == null) {
+            synchronized (TableProvider.class) {
+                if (tableProvider == null) {
+                    tableProvider = new TableProvider();
+                }
+            }
+        }
+        return tableProvider;
+    }
+
+    private synchronized int getTableID() {
+        return this.tableID++;
+    }
+
+    public String createTable(BType constrainedType) {
+        String tableName = TableConstants.TABLE_PREFIX + ((BTableType) constrainedType).getConstrainedType().getName()
+                .toUpperCase() + "_" + getTableID();
+        String sqlStmt = generateCreateTableStatment(tableName, constrainedType);
+        executeStatement(sqlStmt);
+        return tableName;
+    }
+
+    public String insertData(String tableName, BStruct constrainedType) {
+        String sqlStmt = generateInsertDataStatment(tableName, constrainedType);
+        prepareAndExecuteStatement(sqlStmt, constrainedType);
+        return tableName;
+    }
+
+    public void dropTable(String tableName) {
+        String sqlStmt = TableConstants.SQL_DROP + tableName;
+        executeStatement(sqlStmt);
+    }
+
+    public TableIterator createIterator(String tableName, BStructType type) {
+        TableIterator itr;
+        Statement stmt = null;
+        Connection conn = this.getConnection();
+        try {
+            stmt = conn.createStatement();
+            ResultSet rs = stmt.executeQuery(TableConstants.SQL_SELECT + tableName);
```

Comment:
Can we use try-with-resources[1] here?

[1] https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

Reply:
We cannot use try with resource here, since the resources are wrapped into a TableIterator and will be released later once user code retrieves all the data